File: sdk/elixir/lib/bsdkrun/binary.ex
defmodule Bsdkrun.Binary do
  @moduledoc """
  Resolves the `bsdkrun` binary. First match wins; the result is cached in
  `:persistent_term`. Resolution order:

    1. an explicit override set via `set_binary_path/1`,
    2. the `$BSDKRUN_BIN` environment variable,
    3. `bsdkrun` on `$PATH`,
    4. in-repo dev builds relative to this SDK: `<repo>/target/release/bsdkrun`
       then `<repo>/target/debug/bsdkrun`.

  If none exist, `resolve!/0` raises a `Bsdkrun.Error` of kind
  `:binary_not_found` listing everything it looked at.
  """

  alias Bsdkrun.Error

  @override_key {__MODULE__, :override}
  @cache_key {__MODULE__, :resolved}

  @doc """
  Force the SDK to use a specific `bsdkrun` binary, bypassing discovery. Handy
  in tests or when running against a locally built debug binary.
  """
  @spec set_binary_path(String.t()) :: :ok
  def set_binary_path(path) when is_binary(path) do
    :persistent_term.put(@override_key, path)
    :persistent_term.erase(@cache_key)
    :ok
  end

  @doc "Reset cached discovery state and any override (mainly for tests)."
  @spec reset_cache() :: :ok
  def reset_cache do
    :persistent_term.erase(@override_key)
    :persistent_term.erase(@cache_key)
    :ok
  end

  @doc "Resolve (and cache) the path to the `bsdkrun` binary, or raise."
  @spec resolve!() :: String.t()
  def resolve! do
    case :persistent_term.get(@cache_key, nil) do
      nil ->
        searched = candidates()

        case Enum.find_value(searched, &usable/1) do
          nil ->
            raise Error.binary_not_found(searched)

          found ->
            :persistent_term.put(@cache_key, found)
            found
        end

      cached ->
        cached
    end
  end

  # Candidate locations, in priority order.
  defp candidates do
    override = :persistent_term.get(@override_key, nil)
    env = System.get_env("BSDKRUN_BIN")
    on_path = System.find_executable("bsdkrun")

    # This file lives at sdk/elixir/lib/bsdkrun/binary.ex, so the repo root is
    # four levels up. Prefer a release build, fall back to debug.
    repo_root = Path.expand(Path.join(__DIR__, "../../../.."))

    [
      override,
      env,
      on_path,
      Path.join([repo_root, "target", "release", "bsdkrun"]),
      Path.join([repo_root, "target", "debug", "bsdkrun"])
    ]
    |> Enum.reject(&is_nil/1)
  end

  # A path-like candidate must exist on disk; a bare name must be on PATH.
  # Returns the resolved absolute path, or nil if unusable.
  defp usable(candidate) do
    cond do
      String.contains?(candidate, "/") ->
        if File.exists?(candidate), do: candidate, else: nil

      true ->
        System.find_executable(candidate)
    end
  end
end

File: sdk/python/src/bsdkrun/binary.py
"""Locate the ``bsdkrun`` binary on the host.

Resolution order (first match wins, then cached):

1. an explicit override set via :func:`set_binary_path`,
2. the ``$BSDKRUN_BIN`` environment variable,
3. ``bsdkrun`` on ``$PATH``,
4. in-repo dev builds relative to this file:
   ``<repo_root>/target/release/bsdkrun`` then ``.../target/debug/bsdkrun``.
"""

from __future__ import annotations

import os
from pathlib import Path
from shutil import which

from .errors import BinaryNotFound

__all__ = ["set_binary_path", "resolve_binary", "reset_binary_cache"]

_override: str | None = None
_resolved: str | None = None


def set_binary_path(path: str) -> None:
    """Force the SDK to use a specific ``bsdkrun`` binary, bypassing discovery.

    Handy in tests or when running against a locally built debug binary.
    """
    global _override, _resolved
    _override = path
    _resolved = None


def reset_binary_cache() -> None:
    """Reset cached discovery state and any override (mainly for tests)."""
    global _override, _resolved
    _override = None
    _resolved = None


def _repo_root() -> Path:
    # This file lives at sdk/python/src/bsdkrun/binary.py, so the repo root is
    # four directories up (bsdkrun -> src -> python -> sdk -> repo root).
    return Path(__file__).resolve().parents[4]


def _candidates() -> list[str]:
    """Candidate locations, in priority order."""
    out: list[str] = []
    if _override:
        out.append(_override)

    env = os.environ.get("BSDKRUN_BIN")
    if env:
        out.append(env)

    # A `bsdkrun` already on PATH wins over in-repo builds.
    on_path = which("bsdkrun")
    if on_path:
        out.append(on_path)

    repo_root = _repo_root()
    out.append(str(repo_root / "target" / "release" / "bsdkrun"))
    out.append(str(repo_root / "target" / "debug" / "bsdkrun"))
    return out


def resolve_binary() -> str:
    """Resolve (and cache) the path to the ``bsdkrun`` binary.

    Raises :class:`~bsdkrun.errors.BinaryNotFound` if none of the candidate
    locations exist.
    """
    global _resolved
    if _resolved:
        return _resolved

    searched = _candidates()
    for candidate in searched:
        if os.sep in candidate or "/" in candidate:
            if os.path.exists(candidate):
                _resolved = candidate
                return candidate
        else:
            found = which(candidate)
            if found:
                _resolved = candidate
                return candidate
    raise BinaryNotFound(searched)
